pkgs/tools/misc/trackma/default.nix

```
@@ -19,33 +33,49 @@ python3.pkgs.buildPythonApplication rec {
     repo = "trackma";
     rev = "v${version}";
     sha256 = "sha256-drc39ID4WYBQ/L2py57CB5OkQNfRKNigPQW0Lp8GIMc=";
+    fetchSubmodules = true; # for anime-relations submodule
   };
 
-  nativeBuildInputs = lib.optionals withGtk [ wrapGAppsHook ];
+  nativeBuildInputs = lib.optionals withGtk [ wrapGAppsHook ]
+    ++ lib.optionals withQt [ qt5.wrapQtAppsHook ];
 
   buildInputs = lib.optionals withGtk [ glib gobject-introspection gtk3 ];
 
-  propagatedBuildInputs = [ python3.pkgs.urllib3  python3.pkgs.dbus-python ]
-    ++ lib.optionals withGtk [ python3.pkgs.pillow python3.pkgs.pygobject3 python3.pkgs.pycairo ]
+  propagatedBuildInputs = [ python3.pkgs.urllib3 ]
+    ++ lib.optionals withQt [ python3.pkgs.pyqt5 ]
+    ++ lib.optionals withGtk [ python3.pkgs.pycairo ]
     ++ lib.optionals withCurses [ python3.pkgs.urwid ]
-    ++ lib.optionals stdenv.isLinux [ python3.pkgs.pyinotify ];
+    ++ lib.optionals stdenv.isLinux [ python3.pkgs.dbus-python python3.pkgs.pygobject3 python3.pkgs.pyinotify ]
+    ++ lib.optionals (withGtk || withQt) [ python3.pkgs.pillow ];
 
   # broken with gobject-introspection setup hook, see https://github.com/NixOS/nixpkgs/issues/56943
   strictDeps = false;
 
-  dontWrapGApps = true; # prevent double wrapping
+  dontWrapQtApps = true;
+  dontWrapGApps = true;
 
   preFixup = ''
-    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
+    ${lib.optionalString withQt "wrapQtApp $out/bin/trackma-qt"}
+    ${lib.optionalString withGtk "wrapGApp $out/bin/trackma-gtk"}
+  '';
+
+  desktopItemQt = mkDesktopItem "trackma-qt" "Trackma (Qt)" "Trackma Updater (Qt-frontend)" false;
+  desktopItemGTK = mkDesktopItem "trackma-gtk" "Trackma (GTK)" "Trackma Updater (Gtk-frontend)" false;
+  desktopItemCurses = mkDesktopItem "trackma-curses" "Trackma (ncurses)" "Trackma Updater (ncurses frontend)" true;
+
+  postInstall = ''
+    install -Dvm444 $src/$pname/data/icon.png $out/share/pixmaps/$pname.png
+    ${lib.optionalString withQt "install -Dvm444 $desktopItemQt/share/applications/* -t $out/share/applications/"}
+    ${lib.optionalString withGtk "install -Dvm444 $desktopItemGTK/share/applications/* -t $out/share/applications/"}
+    ${lib.optionalString withCurses "install -Dvm444 $desktopItemCurses/share/applications/* -t $out/share/applications/"}
```

_This is wrong in too many levels..._

Well, first you should use the format 

```nix
''
  string
'' ++ lib.optionalString booleanAttribute ''
  optional string
''
```

instead of using `${string interpolation}`. String interpolation is for small expressions; for big expressions a string addition is more readable.

----------------------------

However in this special case you are copying desktop items, there are some useful functionalities:

- `makeDesktopItem` - you are already using it;
- `desktopItems` - it is a list of desktop items, or more precisely a list whose elements are generated by a `makeDesktopItem`
- `copyDesktopItems` - it is a "hook" you put inside `nativeBuildInputs`; it will run an extra `postInstall` that copies the `desktopItems` you declared.

In that particular case you would do something like 

```suggestion
   desktopItems = 
     lib.optional withQt desktopItemQt 
     ++ lib.optional withGTK desktopItemGTK 
     ++ lib.optional withCurses desktopItemCurses;

  postInstall = ''
    install -Dvm444 $src/$pname/data/icon.png $out/share/pixmaps/$pname.png
  '';
```

---

> _This is wrong in too many levels..._

BTW, this code is mostly taken from my (initially not intended to publish) Trackma package, so that's my responsibility as well. And excuse me, but where are these .desktop-related functions documented at all?! There is _no single word_ about desktop items in [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/)! (Are they so well hidden, or am I just looking so bad?) I found this `makeDesktopItem` purely coincidentally, looking at random packages. How are these functions even supposed to be discovered?

---

This is a serious problem on Nixpkgs, to be honest.
Many utilities are created in a pattern recognition basis. Creating scripts and other configuration files is such a repetitive and template-able task, and Nixpkgs evolve in order to create such special packages/functions.

It is a bit harder to document those organic functionalities in a motivated way.

For now they are discovered by grepping the codebase and reading it.